File: 0128.LongestConsecutiveSequence.py
class Solution(object):
    def longestConsecutive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        dic = {}
        ans = 0
        for num in nums:
            if num not in dic:
                left = dic.get(num-1, 0)
                right = dic.get(num+1, 0)
                new = left + right + 1
                ans = max(ans, new)
                dic[num] = dic[num-left] = dic[num+right] = new
            print dic
        return ans

test = Solution()
print test.longestConsecutive([100, 4, 200, 1, 3, 2])

File: 0000-0999/0128.LongestConsecutiveSequence.go
func longestConsecutive(nums []int) int {
    /* O(n) + O(n)
    Map := make(map[int]int)
    cnt := 0
    for _, num := range nums {
        if Map[num] == 0 {
            left, right := Map[num-1], Map[num+1]
            Map[num] = left + right + 1
            Map[num-left] = Map[num]
            Map[num+right] = Map[num]
            cnt = max(cnt, Map[num])
        }
    }*/
    // O(nlgn) + O(1)
    if len(nums) <= 1 {
        return len(nums)
    }
    sort.Ints(nums)
    cnt, tmp := 0, 1
    for i := 1; i < len(nums); i++ {
        if nums[i]-1 == nums[i-1] {
            tmp++
        } else if nums[i] != nums[i-1] { // deal with [0, 1, 1, 2]
            tmp = 1
        }
        cnt = max(cnt, tmp)
    }
    return cnt
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}